src/rust/engine/src/core.rs

```diff
@@ -287,6 +287,8 @@ pub enum Failure {
   /// A Node failed because a filesystem change invalidated it or its inputs.
   /// A root requestor should usually immediately retry their request.
   Invalidated,
+  /// A rule raised an exception about the graph being cyclic.
```

Comment:
Nit: the `graph` crate creates these, rather than `rules`.

Reply:
Thanks!